extensions/amp-a4a/0.1/test/test-a4a-integration.js

```diff
@@ -0,0 +1,187 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+    MockA4AImpl,
+    SIGNATURE_HEADER,
+    TEST_URL,
+} from './utils';
+import {Xhr} from '../../../../src/service/xhr-impl';
+import {createIframePromise} from '../../../../testing/iframe';
+import {
+    data as validCSSAmp,
+} from './testdata/valid_css_at_rules_amp.reserialized';
+import {installDocService} from '../../../../src/service/ampdoc-impl';
+import {adConfig} from '../../../../ads/_config';
+import {a4aRegistry} from '../../../../ads/_a4a-config';
+import {
+    resetScheduledElementForTesting,
+    upgradeOrRegisterElement,
+} from '../../../../src/custom-element';
+import {utf8Encode} from '../../../../src/utils/bytes';
+import * as sinon from 'sinon';
+
+// Integration tests for A4A.  These stub out accesses to the outside world
+// (e.g., XHR requests and interfaces to ad network-specific code), but
+// otherwise test the complete A4A flow, without making assumptions about
+// the structure of that flow.
+
+function expectRenderedInFriendlyIframe(element, srcdoc) {
+  expect(element, 'ad element').to.be.ok;
+  const child = element.querySelector('iframe[srcdoc]');
+  expect(child, 'iframe child').to.be.ok;
+  expect(child.getAttribute('srcdoc')).to.contain.string(srcdoc);
+  const childBody = child.contentDocument.body;
+  expect(childBody, 'body of iframe doc').to.be.ok;
+  [element, child, childBody].forEach(toTest => {
```

Comment:
nitpick: i would avoid the loop here, state them one by one potentially gives you more info when it fails (line number tells you which element is not visible).